addons/udes_stock/models/stock_picking.py

```diff
@@ -739,3 +740,268 @@ def raise_stock_investigation(self, reason, quants, location):
+            values["result_package_id"] = result_package.id
+        if location_dest:
+            values["location_dest_id"] = location_dest.id
+        # Find / prepare move lines by target storage format
+        prepared_mls = self.move_line_ids.prepare_move_lines(
+            product_ids, package, location_src, result_package, location_dest
+        )
+        # Mark moves lines as done / picked
+        for mls, mls_values in prepared_mls.items():
+            mls.mark_as_done(values)
+
+        if result_parent_package and result_package.parent_id != result_parent_package:
+            # Create link to new parent package
+            new_link = PackageHierarchyLink.create(
+                {
+                    "child_id": result_package.id,
+                    "parent_id": result_parent_package.id,
+                    "move_line_ids": [(6, 0, mls.ids)],
+                }
+            )
+        picking = self
+        if self.picking_type_id.u_validate_real_time:
+            picking = self.validate_picking(create_backorder=True)
+
+        return picking
+
+    def validate_picking(self, create_backorder=False, force_validate=False):
+        """ Validates a picking and returns its backorder if any has been created.
+            Will raise an error if create_backorder is False and there are incomplete lines.
+            Optionally will mark as done all lines of the picking if force_validate is set to True.
+        """
+        Picking = self.env["stock.picking"]
+
+        new_picking = Picking.browse()
+        requires_backorder = False
+        mls_to_keep = None
+        with self.statistics() as stats:
+            self.assert_valid_state()
+            # Mark all move lines as done if force validate
+            if force_validate:
+                self.move_line_ids.mark_as_done()
+            # Check if a backorder is required and can be created:
+            # a) Multi users is enabled
+            # b) There are incomplete move lines and create backorder is allowed
+            multi_users_enabled = self.picking_type_id.u_multi_users_enabled
+            if self.move_line_ids.get_lines_incomplete():
+                if multi_users_enabled:
+                    user = self.env.user
+                    mls_to_keep = self.move_line_ids.filtered(
+                        lambda ml: ml.qty_done > 0 and ml.u_done_by == user
+                    )
+                    requires_backorder = True
+                elif not create_backorder:
+                    raise ValidationError(
+                        _(
+                            "Cannot validate %s because there are move lines todo and "
+                            "backorder not allowed"
+                        )
+                        % self.log_name()
+                    )
+                else:
+                    requires_backorder = True
+            if requires_backorder:
+                new_picking = self.backorder_move_lines(mls_to_keep=mls_to_keep)
```

Comment:
This is a drive by comment and not related specifically to this story really but, 
What happens if:
pick has 3 move lines apples, bananas and cherries
user 1 scans apples
user 2 scans bananas

User 1 then drops bananas - it moves the picking bananas and cherries into a backorder. 
Will the user 1 work at the move line level, so validate according to the move line, i.e when they try to drop the move line, the pass the move line id, the picking id is found and then they drop so all is well, or does it use the picking id, which now does not have the banana move lines in it?
Also in this scenario, we may need to consider the user assigned picking id behaviour - essentially the user has been re-assigned, but the start time should remain the same.

Like i say out of scope here - just curious.

Reply:
When multiple users working on the same picking is still not supported, but the idea is to have some code that when the picking is validated and multi users is enabled it should find the backorder in state assigned. We could add that code instead of `assert_valid_state()`.